rmw_zenoh_cpp/src/detail/rmw_client_data.cpp

```diff
@@ -535,12 +511,9 @@ void ClientData::_shutdown()
   }
 
   // Unregister this node from the ROS graph.
-  if (zc_liveliness_token_check(&token_)) {
-    zc_liveliness_undeclare_token(z_move(token_));
-  }
-  if (z_check(z_loan(keyexpr_))) {
-    z_drop(z_move(keyexpr_));
-  }
+  zc_liveliness_undeclare_token(z_move(token_));
+
+  z_drop(z_move(keyexpr_));
```

Comment:
Actually, this is one of the misunderstandings we would like to avoid.

1. There's no way to check an uninitialized zenoh entity.
  ```c
  z_owned_keyexpr_t keyexpr;
  z_check(z_loan(keyexpr))   // This is an undefined behavior.
  ``` 

  A proper way to use `z_check` is to use it on an initialized data.

  ```c
  z_owned_keyexpr_t keyexpr;
  z_keyexpr_from_str(&keyexpr, "foo");
  z_check(z_loan(keyexpr));  // This is okay.
  
  z_owned_keyexpr_t keyexpr;
  z_null(&keyexpr);
  z_check(z_loan(keyexpr)); // This is also fine.
  ```  
2. The check itself checks the gravestone status of zenoh data. It doesn't do functional checks like liveliness token querying over the network. These functions are necessary for zenoh-cpp development yet are easy to misuse. That's why we finally decided to make them internal. See https://github.com/eclipse-zenoh/zenoh-c/pull/605 for details.

3. Regarding the contention issue you mentioned, I assume that the function `ClientData::shutdown` can only be called once a valid ClientData instance has been created. However, if we prefer to relax this restriction, we could introduce a flag to indicate whether the class object has been properly validated. This would help address the issue. What are your thoughts?

Reply:
Thanks for explaining that. Right now `ClientData::shutdown` can be invoked even if the tokens are not properly initialized. So it sounds like we would need to introduce an `initialized_` flag to the implementation.